File: Domains/Base/Workspace/Typescript/Domain/src/allors/domain/base/Session.ts
import { IWorkspace } from './Workspace';
import { ISessionObject, INewSessionObject } from './SessionObject';
import { PushRequest, PushResponse, SyncResponse , ResponseType } from './database';

export interface ISession {
    hasChanges: boolean;

    get(id: string): ISessionObject;

    create(objectTypeName: string): ISessionObject;

    pushRequest(): PushRequest;

    pushResponse(saveResponse: PushResponse): void;
    reset(): void;
}

export class Session implements ISession {
    private static idCounter = 0;

    hasChanges: boolean;

    private sessionObjectById: { [id: string]: ISessionObject; } = {};
    private newSessionObjectById: { [id: string]: INewSessionObject; } = {};

    constructor(private workspace: IWorkspace) {
        this.hasChanges = false;
    }

    get(id: string): ISessionObject {
        if (id === undefined || id === null) {
            return undefined;
        }

        let sessionObject = this.sessionObjectById[id];
        if (sessionObject === undefined) {
            sessionObject = this.newSessionObjectById[id];

            if (sessionObject === undefined) {
                const workspaceObject = this.workspace.get(id);

                const constructor = this.workspace.constructorByName[workspaceObject.objectType.name];
                sessionObject = new constructor();
                sessionObject.session = this;
                sessionObject.workspaceObject = workspaceObject;
                sessionObject.objectType = workspaceObject.objectType;

                this.sessionObjectById[sessionObject.id] = sessionObject;
            }
        }

        return sessionObject;
    }

    create(objectTypeName: string): ISessionObject {
        const constructor = this.workspace.constructorByName[objectTypeName];
        const newSessionObject: INewSessionObject = new constructor();
        newSessionObject.session = this;
        newSessionObject.objectType = this.workspace.metaPopulation.objectTypeByName[objectTypeName];
        newSessionObject.newId = (--Session.idCounter).toString();

        this.newSessionObjectById[newSessionObject.newId] = newSessionObject;

        this.hasChanges = true;

        return newSessionObject;
    }

    reset(): void {
        if (this.newSessionObjectById) {
            Object
                .keys(this.newSessionObjectById)
                .forEach((v) => this.newSessionObjectById[v].reset());
        }

        if (this.sessionObjectById) {
            Object
                .keys(this.sessionObjectById)
                .forEach((v) => this.sessionObjectById[v].reset());
        }

        this.hasChanges = false;
    }

    pushRequest(): PushRequest {
        const data = new PushRequest();
        data.newObjects = [];
        data.objects = [];

        if (this.newSessionObjectById) {
            Object
                .keys(this.newSessionObjectById)
                .forEach((v) => {
                    const newSessionObject = this.newSessionObjectById[v];
                    const objectData = newSessionObject.saveNew();
                    if (objectData !== undefined) {
                        data.newObjects.push(objectData);
                    }
                });
        }

        if (this.sessionObjectById) {
            Object
                .keys(this.sessionObjectById)
                .forEach((v) => {
                    const sessionObject = this.sessionObjectById[v];
                    const objectData = sessionObject.save();
                    if (objectData !== undefined) {
                        data.objects.push(objectData);
                    }
                });
        }

        return data;
    }

    pushResponse(pushResponse: PushResponse): void {
        if (pushResponse.newObjects) {
            Object
                .keys(pushResponse.newObjects)
                .forEach((v) => {
                    const pushResponseNewObject = pushResponse.newObjects[v];
                     const newId = pushResponseNewObject.ni;
                    const id = pushResponseNewObject.i;

                    const newSessionObject = this.newSessionObjectById[newId];

                    const syncResponse: SyncResponse = {
                        responseType: ResponseType.Sync,
                        userSecurityHash: '#', // This should trigger a load on next check
                        objects: [
                            {
                                i: id,
                                v: '',
                                t: newSessionObject.objectType.name,
                                roles: [],
                                methods: []
                            }
                        ]
                    };

                    delete (this.newSessionObjectById[newId]);
                    delete (newSessionObject.newId);

                    this.workspace.sync(syncResponse);
                    const workspaceObject = this.workspace.get(id);
                    newSessionObject.workspaceObject = workspaceObject;

                    this.sessionObjectById[id] = newSessionObject;
                });
        }

        if (Object.getOwnPropertyNames(this.newSessionObjectById).length !== 0) {
            throw new Error('Not all new objects received ids');
        }
    }
}

File: Core/Workspace/CSharp/Domain/Export/Core/Workspace/Session.cs
// <copyright file="Session.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Remote.Push;
    using Allors.Protocol.Remote.Sync;
    using Allors.Workspace.Meta;

    public class Session
    {
        private static long idCounter = 0;

        private readonly Workspace workspace;
        private readonly Dictionary<long, SessionObject> sessionObjectById = new Dictionary<long, SessionObject>();
        private readonly Dictionary<long, SessionObject> newSessionObjectById = new Dictionary<long, SessionObject>();

        public Session(Workspace workspace) => this.workspace = workspace;

        public bool HasChanges => this.newSessionObjectById.Count > 0 || this.sessionObjectById.Values.Any(v => v.HasChanges);

        public SessionObject Get(long id)
        {
            if (!this.sessionObjectById.TryGetValue(id, out var sessionObject))
            {
                if (!this.newSessionObjectById.TryGetValue(id, out sessionObject))
                {
                    var workspaceObject = this.workspace.Get(id);

                    sessionObject = this.workspace.ObjectFactory.Create(this, workspaceObject.ObjectType);

                    sessionObject.WorkspaceObject = workspaceObject;
                    sessionObject.ObjectType = workspaceObject.ObjectType;

                    this.sessionObjectById[workspaceObject.Id] = sessionObject;
                }
            }

            return sessionObject;
        }

        public SessionObject Create(Class @class)
        {
            var newSessionObject = this.workspace.ObjectFactory.Create(this, @class);

            var newId = --Session.idCounter;
            newSessionObject.NewId = newId;
            newSessionObject.ObjectType = @class;

            this.newSessionObjectById[newId] = newSessionObject;

            return newSessionObject;
        }

        public void Reset()
        {
            foreach (var newSessionObject in this.newSessionObjectById.Values)
            {
                newSessionObject.Reset();
            }

            foreach (var sessionObject in this.sessionObjectById.Values)
            {
                sessionObject.Reset();
            }
        }

        public PushRequest PushRequest()
        {
            var data = new PushRequest
            {
                NewObjects = this.newSessionObjectById.Select(v => v.Value.SaveNew()).ToArray(),
                Objects = this.sessionObjectById.Select(v => v.Value.Save()).Where(v => v != null).ToArray(),
            };
            return data;
        }

        public void PushResponse(PushResponse pushResponse)
        {
            if (pushResponse.NewObjects != null && pushResponse.NewObjects.Length > 0)
            {
                foreach (var pushResponseNewObject in pushResponse.NewObjects)
                {
                    var newId = long.Parse(pushResponseNewObject.NI);
                    var id = long.Parse(pushResponseNewObject.I);

                    var newSessionObject = this.newSessionObjectById[newId];

                    var loadResponse = new SyncResponse
                    {
                        UserSecurityHash = "#",
                        // This should trigger a load on next check
                        Objects =
                                                   new[]
                                                       {
                                                           new SyncResponseObject
                                                               {
                                                                   I = id.ToString(),
                                                                   V = string.Empty,
                                                                   T =
                                                                       newSessionObject
                                                                       .ObjectType.Name,
                                                                   Roles = new object[0][],
                                                                   Methods = new string[0][]
                                                               },
                                                       },
                    };

                    this.newSessionObjectById.Remove(newId);
                    newSessionObject.NewId = null;

                    this.workspace.Sync(loadResponse);
                    var workspaceObject = this.workspace.Get(id);
                    newSessionObject.WorkspaceObject = workspaceObject;

                    this.sessionObjectById[id] = newSessionObject;
                }
            }

            if (this.newSessionObjectById != null && this.newSessionObjectById.Count != 0)
            {
                throw new Exception("Not all new objects received ids");
            }
        }
    }
}
